Tests/test-dataframe.lisp

```diff
@@ -0,0 +1,554 @@
+      (case (random 7)
+        (0 nil)
+        (1 (cons t (funcall gen-index*)))
+        (2 (cons (funcall gen-index*) t))
+        (3 (let ((i1 (funcall gen-index))
+                 (i2 (funcall gen-index)))
+             (if (< i1 i2)
+                 (cons i2 i1)
+                 :invalid)))
+        (4 (let ((i1 (funcall gen-index))
+                 (i2 (funcall gen-index*))
+                 (i3 (funcall gen-index*)))
+             (case (random 3)
+               (0 (cons i1 i2))
+               (1 (cons i2 i1))
+               (2 (cons i2 i3)))))
+        (5 (let* ((length (funcall (gen-integer :min 1 :max (length rows))))
+                  (wrong (random length))
+                  (array (make-array length)))
+             (loop for i from 1 below length
+                   do (setf (elt array i)
+                            (funcall gen-index))
+                   finally (setf (elt array wrong)
+                                 (funcall gen-index*))
+                           (return array))))))))
+
+(defun gen-invalid-col-slice (data-frame)
+  (let ((gen-index (gen-valid-col-index data-frame))
+        (gen-index* (gen-invalid-col-index data-frame))
+        (cols (cols data-frame)))
+    (lambda ()
+      (case (random 6)
+        (0 nil)
+        (1 (cons t (funcall gen-index*)))
+        (2 (cons (funcall gen-index*) t))
+        (3 (let ((i1 (funcall gen-index))
+                 (i2 (funcall gen-index)))
+             (cond ((and (integerp i1)
+                         (integerp i2))
+                    (if (< i1 i2)
+                        (cons i2 i1)
+                        :invalid))
+                   ((integerp i1)
+                    (if (< i1 (position i2 cols :test #'string=))
+                        (cons i2 i1)
+                        :invalid))
+                   ((integerp i2)
+                    (if (> i2 (position i1 cols :test #'string=))
+                        (cons i2 i1)
+                        :invalid))
+                   (t
+                    (if (< (position i1 cols :test #'string=)
+                           (position i2 cols :test #'string=))
+                        (cons i2 i1)
+                        :invalid)))))
+        (4 (let ((i1 (funcall gen-index))
+                 (i2 (funcall gen-index*))
+                 (i3 (funcall gen-index*)))
+             (case (random 3)
+               (0 (cons i1 i2))
+               (1 (cons i2 i1))
+               (2 (cons i2 i3)))))
+        (5 (let* ((length (funcall (gen-integer :min 1 :max (length cols))))
+                  (wrong (random length))
+                  (array (make-array length)))
+             (loop for i from 0 below length
+                   do (setf (elt array i)
+                            (funcall gen-index))
+                   finally (setf (elt array wrong)
+                                 (funcall gen-index*))
+                           (return array))))))))
+
+
+
+(defun gen-valid-function (who)
```

Comment:
explain who argument in comment

Reply:
`who` is perhaps a poor word choice. I was thinking something more like this:
```
(defun gen-valid-function (which-function)
  ;; which-function is set to a symbol matching the function name we are
  ;; testing
  (lambda ()
    ;; combination of inputs
    (ecase which-function
      ((map-data-frame-rows add-cols!)
       (lambda (row-index row)
         (declare (ignore row-index))
         (if (> (length row) 2)
             (cons (elt row 0) (elt row 1))
             42)))
      (map-data-frame-cols
       (lambda (col-index col-name value)
         (declare (ignore col-index))
         (cons col-name value)))
      (map-data-frame
       (lambda (row-index row col-index col-name value)
         (declare (ignore row-index col-name))
         (eq (elt row col-index) value))))))
```